### PR TITLE
support GUI mode of 3rd party simulators

### PR DESCRIPTION
**Pull request by ramyam-xilinx**
Reviewer : venkatp

### CODE REVIEW

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 1]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+
+      }
+      temp1 << temp << std::endl;
+    }
+
+    std::remove((simPath + "/simulate.sh").c_str());
+    std::rename((simPath + "/temp.sh").c_str(), (simPath + "/simulate.sh").c_str());
+    systemUtil::makeSystemCall(filePth, systemUtil::systemOperation::PERMISSIONS, "777", boost::lexical_cast<std::string>(__LINE__));
+
+  }
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:165]: Looks like this function is updating the simulation scripts  in driver . Is it not possible to do it in the V++ and provide just command line option from  here

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 2]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
     return buf_size;
   }
 
+  void HwEmShim::writeNewSimulateScript(std::string simPath, std::string simName, std::string simCmd, std::string guiOrBatch, std::string guiEnableOption)
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:127]: Use as much as `const ... &` as you can everywhere.

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 3]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+
+    std::stringstream ss;
+    ss << simPath << "/simulate.sh";
+    std::string filePth = ss.str();
+
+    std::ifstream seek1;
+    std::ofstream temp1;
+    seek1.open(filePth);
+    temp1.open(simPath + "/temp.sh");
+    std::string temp;
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:138]: Ouch... Perhaps you could spend some time studying https://en.cppreference.com/w/ :-)
```c++
std::ifstream seek1 { simPath + "/simulate.sh" };
```
and so on.

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 4]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
     return buf_size;
   }
 
+  void HwEmShim::writeNewSimulateScript(std::string simPath, std::string simName, std::string simCmd, std::string guiOrBatch, std::string guiEnableOption)
+  {
+
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:129]: Useless.

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 5]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+    std::string filePth = ss.str();
+
+    std::ifstream seek1;
+    std::ofstream temp1;
+    seek1.open(filePth);
+    temp1.open(simPath + "/temp.sh");
+    std::string temp;
+
+    while (getline(seek1, temp))
+    {
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:141]: Do not introduce a new coding style. Imitate the previous one...

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 6]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+          if (temp.find(guiOrBatch) != std::string::npos) {
+            temp = std::regex_replace(temp, std::regex(guiOrBatch), guiEnableOption);
+          }
+        }
+
+      }
+      temp1 << temp << std::endl;
+    }
+
+    std::remove((simPath + "/simulate.sh").c_str());
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:161]: Is this an obfuscation contest here? ;-)

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 7]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+    std::rename((simPath + "/temp.sh").c_str(), (simPath + "/simulate.sh").c_str());
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:162]: I see... You are not using `std::filesystem::rename` but the old C API! :-(
Perhaps in 2020 you could use C++17?
And let Jurassic people use older XRT versions?

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx (hunk 6 above)
Comment (keryell) [reply src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:161]: It took me some time to understand that you are using https://en.cppreference.com/w/cpp/io/c/remove and not https://en.cppreference.com/w/cpp/filesystem/rename

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 8]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+    systemUtil::makeSystemCall(filePth, systemUtil::systemOperation::PERMISSIONS, "777", boost::lexical_cast<std::string>(__LINE__));
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:163]: I have no clue about what this line is doing.
It deserves a comment.

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 9]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+
+  }
+
+  void HwEmShim::writeNewSimulateScriptForXcelium(std::string simPath)
+  {
+    //append gui option , if not there already
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:169]: Yeah! A comment!
But fix the formatting.

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 10]
```diff
@@ -123,6 +124,65 @@ namespace xclhwemhal2 {
...
+    std::ofstream temp1;
+    seek1.open(filePth);
+    temp1.open(simPath + "/temp.sh");
+    std::string temp;
+
+    while (getline(seek1, temp))
+    {
+      if (temp.find(simCmd) != std::string::npos)
+      {
+        if (boost::iequals(simName, "xcelium")) {
```
Comment (keryell) [inline src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:144]: Rationale? Do you expect "XcELIuM"?

**Comment by keryell** [comment]
Please add a lot of comments to make the code clearer.
Simplify the code as much as you can: this is open-source code, to be read by a lot of people....
If you have time, take a look at https://isocpp.github.io/CppCoreGuidelines/ and for example https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Rp-what

**Comment by ramyam-xilinx** [comment]
I will recreate a new PULL request with all the suggestions given above.

**Pull request closed without merging**